AppDB/appscale/datastore/datastore_distributed.py

```diff
@@ -558,34 +552,47 @@ def put_entities(self, app, entities, composite_indexes=()):
 
     for encoded_group_key, entity_list in by_group.iteritems():
       group_key = entity_pb.Reference(encoded_group_key)
-      lock = entity_lock.EntityLock(self.zookeeper.handle, [group_key])
 
       txid = self.transaction_manager.create_transaction_id(app, xg=False)
+      self.transaction_manager.set_groups(app, txid, [group_key])
+
+      # Allow the lock to stick around if there is an issue applying the batch.
+      lock = entity_lock.EntityLock(self.zookeeper.handle, [group_key], txid)
       try:
-        with lock:
-          batch = []
-          entity_changes = []
-          for entity in entity_list:
-            prefix = self.get_table_prefix(entity)
-            entity_key = get_entity_key(prefix, entity.key().path())
-
-            current_value = None
-            if current_values[entity_key]:
-              current_value = entity_pb.EntityProto(
-                current_values[entity_key][APP_ENTITY_SCHEMA[0]])
-
-            batch.extend(mutations_for_entity(entity, txid, current_value,
-                                              composite_indexes))
-
-            batch.append({'table': 'group_updates',
-                          'key': bytearray(encoded_group_key),
-                          'last_update': txid})
-
-            entity_changes.append(
-              {'key': entity.key(), 'old': current_value, 'new': entity})
-          self.datastore_batch.batch_mutate(app, batch, entity_changes, txid)
-      finally:
-        self.transaction_manager.delete_transaction_id(app, txid)
+        lock.acquire()
+      except entity_lock.LockTimeout:
+        raise TimeoutError('Unable to acquire entity group lock')
+
+      entity_keys = [
+        get_entity_key(self.get_table_prefix(entity), entity.key().path())
+        for entity in entity_list]
+      current_values = self.datastore_batch.batch_get_entity(
+        dbconstants.APP_ENTITY_TABLE, entity_keys, APP_ENTITY_SCHEMA)
+
+      batch = []
+      entity_changes = []
+      for entity in entity_list:
+        prefix = self.get_table_prefix(entity)
+        entity_key = get_entity_key(prefix, entity.key().path())
+
+        current_value = None
+        if current_values[entity_key]:
+          current_value = entity_pb.EntityProto(
+            current_values[entity_key][APP_ENTITY_SCHEMA[0]])
+
+        batch.extend(mutations_for_entity(entity, txid, current_value,
+                                          composite_indexes))
+
+        batch.append({'table': 'group_updates',
+                      'key': bytearray(encoded_group_key),
+                      'last_update': txid})
+
+        entity_changes.append(
+          {'key': entity.key(), 'old': current_value, 'new': entity})
+      self.datastore_batch.batch_mutate(app, batch, entity_changes, txid)
+
+      lock.release()
```

Comment:
Does timeout suppose to help to resolve issue when lock is acquired but couldn't be released because of failure in between?
Doesn't it make sense to wrap it with try/finally?

Reply:
Are you referring to the LockTimeout when acquiring the lock? If that is thrown, the lock object cleans up after itself.

If the lock is acquired, we don't want the locks to be cleaned up unless everything succeeds. This branch removes the try/finally behavior on the lines you commented because we want the locks to stick around when there is an issue applying the commit. If we wrap it in try/finally, the datastore will always clean up the locks even in cases when a large batch might not be fully committed.